### PR TITLE
Initial implementation of @httpMalformedRequestTests

### PR DESCRIPTION
**Pull request by adamthom-amzn**
*Description of changes:*
Sets up the trait and serialization/deserialization of all of its members. This
trait allows server implementations to generate tests that reject invalid
requests before those requests are bound to a Smithy-generated implementation
class, or before the deserialized request is handed off to application code.

I'm going to keep this in draft mode for feedback while I work on generating tests from it to make sure this is actually useful.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Review by JordonPhillips — CHANGES_REQUESTED** [review]

File: smithy-protocol-test-traits/src/main/resources/META-INF/smithy/smithy.test.smithy [hunk 1]
```diff
@@ -285,3 +285,140 @@ string NonEmptyString
...
+    member: HttpMalformedRequestTestCase
+}
+
+@private
+structure HttpMalformedRequestTestCase {
+    /// The identifier of the test case. This identifier can be used by
+    /// protocol test implementations to filter out unsupported test
+    /// cases by ID, to generate test case names, etc. The provided `id`
+    /// MUST match Smithy's `identifier` ABNF. No two `httpMalformedRequestTests`
+    /// test cases can share the same ID.
```
[the remark this answers is not in the document]
Comment (adamthom-amzn) [reply smithy-protocol-test-traits/src/main/resources/META-INF/smithy/smithy.test.smithy:302]: done

File: smithy-protocol-test-traits/src/main/resources/META-INF/smithy/smithy.test.smithy [hunk 2]
```diff
@@ -285,3 +285,133 @@ string NonEmptyString
...
+    /// to be sent over the wire; if a key or value needs to be
+    /// percent-encoded, then it MUST appear percent-encoded in this list.
+    ///
+    /// A serialized HTTP request is not in compliance with the protocol
+    /// if any query string parameter defined in `queryParams` is not
+    /// defined in the request or if the value of a query string parameter
+    /// in the request differs from the expected value.
+    ///
+    /// `queryParams` applies no constraints on additional query parameters.
+    queryParams: StringList,
```
Comment (gosar) [inline smithy-protocol-test-traits/src/main/resources/META-INF/smithy/smithy.test.smithy:391]: no queryParams in response

File: smithy-protocol-test-traits/src/test/resources/software/amazon/smithy/protocoltests/traits/errorfiles/all-malformed-request-features.smithy [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+            method: "POST",
+            uri: "/",
+            queryParams: ["foo=baz"]
+        },
+        response: {
+            code: 400,
+            headers: {"X-Foo": "baz"},
+            body: """
+            {
+                "message" : "Invalid JSON"
```
Comment (gosar) [inline smithy-protocol-test-traits/src/test/resources/software/amazon/smithy/protocoltests/traits/errorfiles/all-malformed-request-features.smithy:52]: Is this the same test as above but checking body v/s bodyMessageRegex? The regex has ": .*" but it doesn't here...

File: smithy-protocol-test-traits/src/test/resources/software/amazon/smithy/protocoltests/traits/errorfiles/all-malformed-request-features.smithy [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+
+use smithy.test#httpMalformedRequestTests
+
+@trait
+@protocolDefinition
+structure testProtocol {}
+
+@trait
+@authDefinition
+structure testScheme {}
```
Comment (gosar) [inline smithy-protocol-test-traits/src/test/resources/software/amazon/smithy/protocoltests/traits/errorfiles/all-malformed-request-features.smithy:11]: why is this needed?

File: smithy-protocol-test-traits/src/test/resources/software/amazon/smithy/protocoltests/traits/errorfiles/all-malformed-request-features.smithy (hunk 3 above)
Comment (adamthom-amzn) [reply smithy-protocol-test-traits/src/test/resources/software/amazon/smithy/protocoltests/traits/errorfiles/all-malformed-request-features.smithy:52]: you can either specify body (exact match) or a regex for the message. this test is simply "can we deserialize all of these fields"

File: smithy-protocol-test-traits/src/test/resources/software/amazon/smithy/protocoltests/traits/errorfiles/all-malformed-request-features.smithy (hunk 4 above)
Comment (adamthom-amzn) [reply smithy-protocol-test-traits/src/test/resources/software/amazon/smithy/protocoltests/traits/errorfiles/all-malformed-request-features.smithy:11]: it's not

**Review by gosar — APPROVED** [review]

**Review by JordonPhillips — APPROVED** [review]

File: smithy-protocol-test-traits/src/main/resources/META-INF/smithy/smithy.test.smithy [hunk 5]
```diff
@@ -285,3 +285,119 @@ string NonEmptyString
...
+
+@private
+structure HttpMalformedRequestTestCase {
+    /// The identifier of the test case. This identifier can be used by
+    /// protocol test implementations to filter out unsupported test
+    /// cases by ID, to generate test case names, etc. The provided `id`
+    /// MUST match Smithy's `identifier` ABNF. No two `httpMalformedRequestTests`
+    /// test cases can share the same ID.
+    @required
+    @pattern("[A-Za-z_][A-Za-z0-9_]+")
```
Comment (mtdowling) [inline smithy-protocol-test-traits/src/main/resources/META-INF/smithy/smithy.test.smithy:304]: You need to anchor this

File: smithy-protocol-test-traits/src/main/resources/META-INF/smithy/smithy.test.smithy [hunk 6]
```diff
@@ -285,3 +285,119 @@ string NonEmptyString
...
+
+    /// The expected response.
+    @required
+    response: HttpMalformedResponseDefinition,
+
+    /// A description of the test and what is being asserted.
+    documentation: String,
+
+    /// Applies a list of tags to the test.
+    tags: NonEmptyStringList,
```
Comment (mtdowling) [inline smithy-protocol-test-traits/src/main/resources/META-INF/smithy/smithy.test.smithy:324]: Is there a client/server divide here too? If so, then it's probably good to also add the same client/server controls that were added to the protocol tests here.

File: smithy-protocol-test-traits/src/main/resources/META-INF/smithy/smithy.test.smithy (hunk 5 above)
Comment (adamthom-amzn) [reply smithy-protocol-test-traits/src/main/resources/META-INF/smithy/smithy.test.smithy:304]: That's what I get for copy and pasting HttpRequestTestCase

File: smithy-protocol-test-traits/src/main/resources/META-INF/smithy/smithy.test.smithy (hunk 6 above)
Comment (adamthom-amzn) [reply smithy-protocol-test-traits/src/main/resources/META-INF/smithy/smithy.test.smithy:324]: I do not see there being an application for malformed request testing on clients, which is why I omitted it.

File: smithy-protocol-test-traits/src/main/resources/META-INF/smithy/smithy.test.smithy (hunk 5 above)
Comment (mtdowling) [reply smithy-protocol-test-traits/src/main/resources/META-INF/smithy/smithy.test.smithy:304]: Hah! We need to anchor that one too!

Comment (adamthom-amzn) [reply smithy-protocol-test-traits/src/main/resources/META-INF/smithy/smithy.test.smithy]: I added it as another commit to this PR.

**Review by JordonPhillips — APPROVED** [review]
